File: app/src/main/java/app/opass/ccip/adapter/AnnouncementAdapter.java
package app.opass.ccip.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import app.opass.ccip.R;
import app.opass.ccip.model.Announcement;
import app.opass.ccip.util.LocaleUtil;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import app.opass.ccip.model.Announcement;
import app.opass.ccip.util.LocaleUtil;

public class AnnouncementAdapter extends RecyclerView.Adapter<ViewHolder> {

    private Context mContext;
    private final TypedValue mTypedValue = new TypedValue();
    private int mBackground;
    private List<Announcement> announcementList;

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private TextView msg, time;

        public ViewHolder(View itemView) {
            super(itemView);
            msg = (TextView) itemView.findViewById(R.id.invalid_token_msg);
            time = (TextView) itemView.findViewById(R.id.time);
        }
    }

    public AnnouncementAdapter(Context context, List<Announcement> announcementList) {
        mContext = context;
        this.announcementList = announcementList;

        context.getTheme().resolveAttribute(R.attr.selectableItemBackground, mTypedValue, true);
        mBackground = mTypedValue.resourceId;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_announcement, parent, false);

        itemView.setBackgroundResource(mBackground);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, final int position) {
        final ViewHolder holder = ((ViewHolder) viewHolder);
        SimpleDateFormat sdf = new SimpleDateFormat("MM/dd HH:mm");
        final Announcement announcement = announcementList.get(position);

        if (LocaleUtil.getCurrentLocale(mContext).toString().startsWith(Locale.TAIWAN.toString())) {
            holder.msg.setText(announcement.getMsgZh());
        } else {
            holder.msg.setText(announcement.getMsgEn());
        }
        holder.time.setText(sdf.format(new Date(announcement.getDatetime() * 1000L)));
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (!announcement.getUri().isEmpty()) {
                    mContext.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(announcement.getUri())));
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return announcementList.size();
    }
}

File: app/src/main/java/app/opass/ccip/ui/announcement/AnnouncementAdapter.kt
package app.opass.ccip.ui.announcement

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import app.opass.ccip.R
import app.opass.ccip.model.Announcement
import app.opass.ccip.util.LocaleUtil
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class AnnouncementAdapter(private val mContext: Context, private val announcementList: List<Announcement>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val mTypedValue = TypedValue()
    private val mBackground = mTypedValue.resourceId

    init {
        mContext.theme.resolveAttribute(android.R.attr.selectableItemBackground, mTypedValue, true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_announcement, parent, false)

        itemView.setBackgroundResource(mBackground)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val holder = viewHolder as ViewHolder
        val sdf = SimpleDateFormat("MM/dd HH:mm")
        val (datetime, msgEn, msgZh, uri) = announcementList[position]

        if (LocaleUtil.getCurrentLocale(mContext).toString().startsWith(Locale.TAIWAN.toString())) {
            holder.msg.text = msgZh
        } else {
            holder.msg.text = msgEn
        }
        holder.time.text = sdf.format(Date(datetime * 1000L))
        holder.itemView.setOnClickListener {
            if (!uri.isEmpty()) {
                mContext.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(uri)))
            }
        }
    }

    override fun getItemCount(): Int {
        return announcementList.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val msg: TextView = itemView.findViewById(R.id.invalid_token_msg)
        val time: TextView = itemView.findViewById(R.id.time)
    }
}
